### PR TITLE
ARTEMIS-4372 Implement Pico-cli and script auto-complete

### PR DESCRIPTION
**Pull request by clebertsuconic**

### CODE REVIEW

File: artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/AutoCompletion.java
Comment (jbertram) [inline artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/AutoCompletion.java:34]: Reference to "check-leak" here seems wrong.

**Comment by clebertsuconic** [comment]
https://github.com/apache/activemq-artemis/assets/750514/2e094597-8172-44e6-9965-e88766c396e2

File: docs/user-manual/en/using-server.md
Comment (gemmellr) [inline docs/user-manual/en/using-server.md:61]: Sentence fragment, seems to end mid:

---
Also, this drop-in about the shell usage also doesnt really fit with the rest of the existing information on the page below that all assumes you are **not** using it. It could do with some elaboration / description linking and contrasting the different usages so the two routes are clear, and some new user actually following this page linearly isnt met with a sequence that essentially says 'do foo' followed by 'do this incompatible thing that requires you didnt do the earlier thing we already said'.

File: artemis-cli/src/main/java/org/apache/activemq/artemis/cli/Artemis.java
Comment (clebertsuconic) [reply artemis-cli/src/main/java/org/apache/activemq/artemis/cli/Artemis.java:213]: we shouldn't really add anything to do the CommandLine that's not a Runnable or an Action. But I will add a check.

That should prevent future devs from making a mistake.

File: artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/Create.java
Comment (clebertsuconic) [reply artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/Create.java:42]: we could use import static.. which I don't like much.

But I can make an exception here

File: docs/user-manual/en/using-server.md
Comment (gemmellr) [reply docs/user-manual/en/using-server.md:61]: A note on how to exit the shell may also be in order, some people wont actually know.

File: artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/Create.java
Comment (gemmellr) [reply artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/Create.java:42]: Dont think it should need to be static, its just a nested interface/class, regular import of the full name should work.

File: artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/user/UserGroup.java
Comment (gemmellr) [inline artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/user/UserGroup.java:23]: These group examples like "(example ./artemis user list)" show up in the shell help output, where it is actually _not_ what should be done. Not sure if it could cater to shell vs not-shell usage mode, but might be nice if it was possible.

Comment (clebertsuconic) [reply artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/user/UserGroup.java]: I think I can specify the description on the API and use the isShell that I implemented. will take a look.

Comment (clebertsuconic) [reply artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/user/UserGroup.java]: My options here are limited. I could create a different DataGroup for ShellDataGroup, or I can make a more generic text.

The other option would be some horrible hacking into the internal API.. which would be even worse. So I will go for the more generic text.

Comment (clebertsuconic) [reply artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/user/UserGroup.java]: I think I have solved this..

@gemmellr  but I will let you decide if this one could be resolved or not

**Comment by clebertsuconic** [comment]
I have added one more review from myself. When using user/password the system would ask me for the user/password all the time. I'm now saving the last successful user/password between executions, adding a connect and disconnect that's only available through the shell (what makes the shell more useful since it won't ask for user/password every time when using security)

**Comment by clebertsuconic** [comment]
I will finish the other 2 items tomorrow. there's one item I didn't close (resolve) from our conversation here because I'm not sure it's the best choice yet.. will look at it again tomorrow.

File: artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/messages/DestAbstract.java
Comment (gemmellr) [inline artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/messages/DestAbstract.java:40]: The non-deprecated option name being first would read better (unless the cli is reverse-ordering the options)

File: docs/user-manual/en/using-server.md
Comment (gemmellr) [reply docs/user-manual/en/using-server.md:61]: unresolved as the docs didnt seem to change

**Comment by gemmellr** [comment]
> I have added one more review from myself. When using user/password the system would ask me for the user/password all the time. I'm now saving the last successful user/password between executions, adding a connect and disconnect that's only available through the shell (what makes the shell more useful since it won't ask for user/password every time when using security)

I think the ability to set and clear the URI + credentials used for subsequent commands in the shell is good. I'm less clear on it doing it silently in the background based on every previous command.

At least one of the additions you made was around it handling the URIs crossing between the different clients, presumably as the URI is saved but the protocol value is not, probably how you hit that, the client used will change between commands in some cases, whilst the URI + credentials etc do not even though the URIs are client-specific.

File: artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/messages/ConnectionAbstract.java
Comment (clebertsuconic) [reply artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/messages/ConnectionAbstract.java:179]: the connection is already called on line 177

File: docs/user-manual/en/using-server.md
Comment (clebertsuconic) [reply docs/user-manual/en/using-server.md]: I removed the line.. and I'm adding a chapter for the CLI.

File: artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/ActionAbstract.java
Comment (clebertsuconic) [reply artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/ActionAbstract.java:213]: the Shell will only call the Runnable. So this to interface back to the execute method.

File: artemis-cli/src/main/java/org/apache/activemq/artemis/cli/Artemis.java
Comment (gemmellr) [inline artemis-cli/src/main/java/org/apache/activemq/artemis/cli/Artemis.java:93]: Is this a leftover?

File: artemis-distribution/src/main/resources/licenses/bin/LICENSE
Comment (gemmellr) [inline artemis-distribution/src/main/resources/licenses/bin/LICENSE:293]: Referenced file still needs added.

File: docs/user-manual/en/using-server.md
Comment (gemmellr) [inline docs/user-manual/en/using-server.md:56]: an -> a

I would put the sentence at the start of the existing paragraph below about broker instances.

File: artemis-cli/src/main/java/org/apache/activemq/artemis/cli/Artemis.java
Comment (clebertsuconic) [reply artemis-cli/src/main/java/org/apache/activemq/artemis/cli/Artemis.java:93]: yeah!! lol thank you

File: artemis-distribution/src/main/resources/licenses/bin/LICENSE
Comment (clebertsuconic) [reply artemis-distribution/src/main/resources/licenses/bin/LICENSE:293]: It was added here but apparently .txt is on the gitignore and I missed adding it. thank you.

File: artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/messages/DestAbstract.java
Comment (clebertsuconic) [reply artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/messages/DestAbstract.java:40]: I'm using hidden now...


and I opened a PR and bug on pico to fix the auto-completion on the shell:

https://github.com/remkop/picocli/pull/2075

**Comment by clebertsuconic** [comment]
for my future reference. I opened an issue and Pull Request on Pico: https://github.com/remkop/picocli/pull/2075

the jshell auto completion is showing hidden values. the PR sent would fix it but I'm waiting for a best way to merge this upstream in Pico. hidden is the way to go to "deprecate" stuff.

**Comment by clebertsuconic** [comment]
Only thing left on my todo is the JDK parameters and testing with a newer version. will do it before the end of today.

File: artemis-distribution/src/main/resources/bin/artemis
Comment (clebertsuconic) [reply artemis-distribution/src/main/resources/bin/artemis:96]: hmmmmmmm... I don't need these with JDK17.

@gnodet it would be so nice to get your hep here? :)

Comment (clebertsuconic) [reply artemis-distribution/src/main/resources/bin/artemis:96]: @gnodet I"m adding jline3 into artemis, integrated with pico-CLI. This PR is taking an example from the jline3 along with some examples I saw on pico-cli .. and I had to include these with JDK-11, while these are not needed in 17.


I'm confused what I need to do to integrate this. can you help us here?

Comment (gemmellr) [reply artemis-distribution/src/main/resources/bin/artemis:96]: Its not actually surprising that behaviour here might differ on different JDKs, thats exactly why I wondered if you tried it on the newer ones, especially with JDK21 almost done (RC1 due in 10 days, later RC2 if needed; JDK17 didnt get an RC2)...though I was more concerned it may need other config, than not need any, we will need to use what works for them all.

File: artemis-cli/src/main/resources/org/apache/activemq/artemis/cli/commands/bin/artemis
Comment (gnodet) [inline artemis-cli/src/main/resources/org/apache/activemq/artemis/cli/commands/bin/artemis:107]: Adding the jansi library will remove the need for those additional JVM options.  Those are only required when using the exec provider, which won't be used if you add jansi to support windows.

File: artemis-distribution/src/main/resources/bin/artemis
Comment (gnodet) [reply artemis-distribution/src/main/resources/bin/artemis:96]: See above, adding jansi will remove the need for those options.

**Review by gnodet — CHANGES_REQUESTED** [review]
You need to add jansi library, which is an optional dependency of JLine for better support.

File: artemis-cli/pom.xml
Comment (clebertsuconic) [reply artemis-cli/pom.xml:137]: @gnodet thank you so much.. that worked like a charm!


@gemmellr  ^^^ C/C

Comment (clebertsuconic) [reply artemis-cli/pom.xml:137]: @gnodet thank you so much for your help here.

**Comment by clebertsuconic** [comment]
@gemmellr I am merging this... further changes can be sent into main after this.


Thanks for your help on this

File: artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/messages/DestAbstract.java
Comment (gemmellr) [inline artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/messages/DestAbstract.java:151]: If both were somehow set, I would expect the new one to always win...here the old one will win.

File: artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/messages/perf/PerfCommand.java
Comment (gemmellr) [inline artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/messages/perf/PerfCommand.java:77]: If both were somehow set, I would expect the new one to always win...here the old one will win.

File: artemis-cli/src/main/java/org/apache/activemq/artemis/cli/Artemis.java
Comment (gemmellr) [inline artemis-cli/src/main/java/org/apache/activemq/artemis/cli/Artemis.java:205]: Good chance the assert also wont happen unless people turn on assertions. Seems like a null check would have been simpler. Though it will at least still NPE later, when failing to throw the IAE.

Comment (clebertsuconic) [reply artemis-cli/src/main/java/org/apache/activemq/artemis/cli/Artemis.java:205]: this is never supposed to happen. 
Only chance userObject is null is if you set the userObject wrongly during development. and picocli actually validate that already.

File: artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/messages/ConnectionAbstract.java
Comment (clebertsuconic) [reply artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/messages/ConnectionAbstract.java:289]: not really in both cases.. This is the exception of the exception. IN case the input user/password does not succeed, if this does not work this will simply not save the password here.

Comment (clebertsuconic) [reply artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/messages/ConnectionAbstract.java:289]: I see what you mean now.

File: artemis-cli/src/main/java/org/apache/activemq/artemis/cli/Artemis.java
Comment (gemmellr) [reply artemis-cli/src/main/java/org/apache/activemq/artemis/cli/Artemis.java:205]: Fair enough. You didnt note the last bit before, just that 'we shouldnt add things that arent Runnable or Action', then you added the fallback check which would NPE if it was null, which I then noted.  I'd still have just added a simple Objects.nonNull("oh-dear developer-error", userObject) rather than the multi line assert+comment that quite posibly still wont do anything and allow the raw NPE to occur.

**Pull request merged**
